### PR TITLE
feat(contents_text): Add `contents_*()` helper functions

### LINKED ISSUE #167: Additional generics to convert turns and turn contents into other text-based formats
Currently the `Turn@text` getter is used to convert a turn into text by calling `contents_text()` on the elements in the turn's `@contents`. The text representation, however, loses critical information, such as hiding the images attached to the contents.

I think it's reasonable for this to happen with `Turn@text`, but for more complicated uses it'd be helpful to have a way to also get an html or markdown representation of the Turns.

Rather than increasing the number of properties on `Turn`, I'd like to propose that `contents_text()` becomes an exported function and is joined by `contents_markdown()` and `contents_html()` (possibly also supported by a extending an underlying, possibly exported, abstraction of `contents_transform()`).

Ideally, these functions could be defined for the content classes `ContentText`, `ContentInlineImage`, etc. as well as for `Turn`, such that a user could call `lapply(chat$get_turns(), contents_html)` or equally `contents_html(chat$get_turns()[[1]])`.

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #167

Adds `contents_text()`, `contents_markdown()` and `contents_html()` generics that can be applied to `Turn` and `Content` objects.

``` r
turns <- list(
  user_turn("What's this image?", content_image_url("https://placehold.co/200x200")),
  Turn("assistant", "It's a placeholder image.")
)

lapply(turns, contents_text)
#> [[1]]
#> [1] "What's this image?"
#> 
#> [[2]]
#> [1] "It's a placeholder image."
lapply(turns, contents_markdown)
#> [[1]]
#> [1] "What's this image?\n\n![](https://placehold.co/200x200)"
#> 
#> [[2]]
#> [1] "It's a placeholder image."
contents_html(turns[[1]])
#> [1] "<p>What's this image?</p>\n\n<img src=\"https://placehold.co/200x200\">"
```

### CODE REVIEW

File: R/turns.R [hunk 1]
```diff
@@ -44,9 +44,7 @@ Turn <- new_class(
     ),
     text = new_property(
       class = class_character,
-      getter = function(self) {
-        paste0(unlist(lapply(self@contents, contents_text)), collapse = "")
-      }
+      getter = function(self) contents_text(self)
```
Comment (hadley) [inline R/turns.R:47]: You didn't change anything to make this work, right? You just noticed that we can now simplify the implementation?

File: _pkgdown.yml [hunk 2]
```diff
@@ -38,3 +38,7 @@ reference:
   - Provider
   - Content
   - Chat
+
+- title: Utilities
+  contents:
+  - contents_text
```
Comment (hadley) [inline _pkgdown.yml:44]: Missing trailing nl

File: R/content.R [hunk 3]
```diff
@@ -1,6 +1,47 @@
 #' @include utils-S7.R
 NULL
 
+#' Format contents into a textual representation.
```
Comment (hadley) [inline R/content.R:4]:
```suggestion
#' Format contents into a textual representation
```

File: R/content.R [hunk 4]
```diff
@@ -1,6 +1,47 @@
...
+#' )
+#'
+#' lapply(turns, contents_text)
+#' lapply(turns, contents_markdown)
+#' if (rlang::is_installed("commonmark")) {
+#'   contents_html(turns[[1]])
+#' }
+#' 
+#' @param content The [Turn] or [Content] object to be converted into text.
+#' @return A character string of text, markdown or HTML.
```
Comment (hadley) [inline R/content.R:32]:
```suggestion
#' @return A string containing text, markdown, or HTML.
```

File: R/content.R [hunk 5]
```diff
@@ -1,6 +1,47 @@
...
+#'   Turn("assistant", "It's a placeholder image.")
+#' )
+#'
+#' lapply(turns, contents_text)
+#' lapply(turns, contents_markdown)
+#' if (rlang::is_installed("commonmark")) {
+#'   contents_html(turns[[1]])
+#' }
+#' 
+#' @param content The [Turn] or [Content] object to be converted into text.
```
Comment (hadley) [inline R/content.R:31]: Can you pass turns?

**Review by hadley — COMMENTED** [review]
This seems like a reasonable idea and implementation to me. You just need to fix the build failure and add some basic tests.

File: R/turns.R (hunk 1 above)
Comment (gadenbuie) [reply R/turns.R:47]: Right, I moved the implementation that was here into `method(contents_text, Turn)`.

File: R/content.R (hunk 5 above)
Comment (gadenbuie) [reply R/content.R:31]: I just added a method for `contents_markdown()` for `Chat` instances. I see a lot of utility in that particular output format and less utility in the HTML or text variants. That said I'd be happy to add those if you want.

``` r
turns <- list(
  user_turn("What's this image?", content_image_url("https://placehold.co/200x200")),
  Turn("assistant", "It's a placeholder image.")
)

chat <- Chat$new(Provider("https://example.com"), turns = turns)
cat(contents_markdown(chat))
#> ## User
#> 
#> What's this image?
#> 
#> ![](https://placehold.co/200x200)
#> 
#> ## Assistant
#> 
#> It's a placeholder image.
```

Comment (gadenbuie) [reply R/content.R:31]: Related: is there a way to create a `chat_echo()` (or maybe `chat_repeater()`) chat instance that just returns the user input and doesn't call any APIs? Carson made an [example like this for Shiny](https://shiny.posit.co/py/components/display-messages/chat/) and I've found it to be very useful for debugging and examples when I'm not ready for real model usage yet.

**Comment by gadenbuie** [comment]
@hadley this is ready for another look. Sorry for the commit noise, I only just now realized I wasn't able to run checks locally because I had out-of-date credentials. Of course when I finally got to ✅ in the last commit there was an unrelated CI failure and I can't request a rerun.

File: R/content.R (hunk 5 above)
Comment (hadley) [reply R/content.R:34]: I think that would be pretty straightforward with a new provider.

**Comment by cpsievert** [comment]
I love this idea in general, but also had a couple thoughts on this PR:

1. It'd be nice if these content extractors had the ability to get "additional" content (i.e., the stuff you get with `$chat(echo="all")`), mainly for debugging (but also educational) purposes.
2. It feels like the main way people would want to use this is to generate an md/html _file_ of the whole conversation? For that reason, instead of generics, I think for chatlas I might do something like `chat.export(filename="output.html", content="all")`, which might internally call a new `turn.extract_content("all")`.

**Comment by hadley** [comment]
I'm going to merge this because it's useful today, but @cpsievert's extension ideas definitely sound good to me (and might be worth tracking as issues if either of you wants to add)

**Pull request merged**